### DISCUSSION TITLE
Set a cell comment to RTL

### DISCUSSION THREAD

**Discussion by berberon** [PHPOffice/PhpSpreadsheet#4004] [General]
Is it possible to set a Cell Comment to RTL?
Please note I'm not talking about the cell's content but a comment added by createTextRun().

Thanks

**Comment by oleibman** [comment]
Is there more to your question than you are asking? I can certainly add an RTL text run to a comment, and it will display RTL. Is your problem that the overall comment is aligned LTR? Or are you having trouble with mixed RTL and LTR? Or some other problem?

**Comment by berberon** [comment]
Hi

The comments are always both aligned LTR and aligned left.
I have : 

$objPhpSpreadsheet->getActiveSheet()->getStyle("A1")->getAlignment()->setReadorder(Alignment::READORDER_RTL);

but this does not affect the comment.

Can you share an example of how i can get the comment to be both RTL and Aligned right?

**Comment by oleibman** [comment]
There is not currently a way to right-align the comment. I think I see how it might be done _on write_; however, reading and recognizing that a comment is right-aligned looks like it would be considerably more difficult. I am investigating.

**Comment by oleibman** [comment]
Should be fixed by PR #4006.

**Comment by berberon** [comment]
Should i do anything in my code to get this to work?

**Reply by oleibman** [reply]
It is included with Release 2.1.0. If you have installed that release, I don't think you need to make any code changes for it to work.

**Reply by berberon** [reply]
This does not seem to be working.
I can get the text in the comment to align right but the RTL is not working.
Looking at the button in the menu while the note is open... it's always on "context" and not on RTL

```
	$objPHPExcel->getActiveSheet()->getComment('A1')->getText()->createTextRun("\r\n");
	$objPHPExcel->getActiveSheet()->getComment('A1')->getText()->createTextRun($SettingsStamp);
	$objPHPExcel->getActiveSheet()->getComment('A1')->setAlignment(Alignment::READORDER_RTL);
	$objPHPExcel->getActiveSheet()->getComment('A1')->setWidth("390px");
	$objPHPExcel->getActiveSheet()->getComment('A1')->setHeight("1000px");
```



![QFCU5SLsj3](https://github.com/PHPOffice/PhpSpreadsheet/assets/6428354/9621ffe2-515a-4f9b-907d-525c9151690a)

versions : * 2.1.0
released : 2024-05-11, 3 weeks ago

**Reply by oleibman** [reply]
The alignment property belongs to the entire comment, so, yes, setting it twice will cause the second one to overlay the first. I am not sure what it is you are trying to accomplish. Can you upload a spreadsheet created in Excel (or at least a picture) that has a comment displayed the way you want?

**Reply by berberon** [reply]
[PricezBINew_רשימת_מוצרים_1_2024-06-04-21-07.xlsx](https://github.com/user-attachments/files/15568640/PricezBINew_._._1_2024-06-04-21-07.xlsx)


Check out the note in Cell A1 in the attached excel.
I want it to be both Aligned right and RTL.

I can do it manually but not via phpspreadsheet

**Reply by berberon** [reply]
I think i found a workaround....

Instead of using 

`$objPHPExcel->getActiveSheet()->getComment('A1')->setAlignment(Alignment::READORDER_RTL);`

you can do add the RTL char "\u{202E}":

`$objPHPExcel->getActiveSheet()->getComment('A1')->getText()->createTextRun("\u{202E}" . $SettingsStamp);`


This is a work around but it seems to work.

Is there a way to do it without this workaround?

**Comment by oleibman** [comment]
In the spreadsheet which you uploaded, the text is left-aligned, but you said you wanted it right-aligned. As for the RTL, the colons which are supposed to terminate some of the lines are definitely misplaced. And I'm not sure, but I think the percent signs might also be misplaced. You might indeed need Right-to-Left Override (202e) (and this should be paired with Pop Directional Format 202c) or Right-to-Left Mark (202f - probably preferable to RLO for single characters) to ensure that they display correctly. A better choice might be Right-to-Left Isolate (2067) or First Strong Isolate (2068) paired with Pop Directional Isolate (2069), but, the last time I checked, Excel didn't handle those properly, and I just confirmed that it still doesn't.

At any rate, placing the trailing colon is a problem for most apps which handle bidirectional text, including text editors and browsers. They all need the control characters or special tags to line things up the way you want them. Adding them isn't really a workaround; it's just the way the Unicode bidirectional algorithm works.

At any rate, I am working on a solution which might be a little easier for you. It puts the trailing colons in the right place:
![image](https://github.com/PHPOffice/PhpSpreadsheet/assets/10341515/9da0ce9b-bb9a-42f0-a19e-e9a9309f3a20)
And it puts the percent signs in what I think is the right place:
![image](https://github.com/PHPOffice/PhpSpreadsheet/assets/10341515/fb81cbec-1360-46ab-a51d-2c418b8a7c76)
But the following lines look really messed up to me:
![image](https://github.com/PHPOffice/PhpSpreadsheet/assets/10341515/2408c1a9-067e-4f4b-960f-ecd65a06f4db)
To get those lines displayed correctly, I think you have no choice but to use the Unicode directional characters.
Here's another example, in text rather than pictures:
```
יָנוּאָר 23,‏ 1954.‏
```
To get that to work correctly, I needed to add an LRM (200e) after the comma and after the period. You'll probably need to do something similar with the "messed up" lines above.

**Reply by berberon** [reply]
The excel i uploaded was to show the problem.
I can set the A1 Note to aligh left but if i don't manually add the "\u{202E}" before the text in the note it will be LTR and not RTL.

In excel, I have the option to choose (Force) RTL by using this menu : 

![QFCU5SLsj3](https://github.com/PHPOffice/PhpSpreadsheet/assets/6428354/09e056c6-249b-451a-b697-97d1bba2ef68)


What i don't have, is the option in phpspreadsheet to force this menu to be RTL... i'ts always in Context mode


Here is the same excel with the "\u{202E}". it's perfect.

[Uploading Good.xlsx…]()

I rather have a code option to force RTL than add "\u{202E}" everywhere.

P.S.

The % at the begining is good, it's not a problem. It's suposed to be there.

**Reply by oleibman** [reply]
It looks like your latest upload may not have succeeded.

**Reply by berberon** [reply]
[Good.xlsx](https://github.com/user-attachments/files/15582799/Good.xlsx)


Sorry about that, trying again.

**Reply by berberon** [reply]
Just to make sure we see the same thing... this is perfect but only works if i force RTL using "\u{202E}"



![J9WibKJbAT](https://github.com/PHPOffice/PhpSpreadsheet/assets/6428354/332abad9-c5cf-4361-989a-f812bbfb4fa6)

**Comment by oleibman** [comment]
Thank you, I have downloaded and what you show matches what I see. I will work with that. However, I'm still not sure about the 2 "messed up" lines I show above - the lines with `X` and `Y` in them. They show up below your screenshot, but they still look messed up to me. Are they formatted correctly or not? If not, can you describe what they should look like ideally?

**Reply by berberon** [reply]
You are correct, this line is the only one that is not ok : 

`X יחידות ב-Y ₪ = 0`

The char that messes it up is "₪".
I can simply replace it with 'ש"ח' which means NIS instead and it will be good.

**Comment by oleibman** [comment]
See PR #4065.
